### ISSUE TITLE
Internal parameter in match types doesn't work as expected

### ISSUE THREAD

**Issue by heaven-born** [lampepfl/dotty#12044]
## Compiler version

3.0.0-RC1

## Minimized code

```Scala
trait A
trait B

trait AA
trait BB

type Hpr[X] = X match
    case A & B & t => AA & BB & t

summon[(AA & BB & Int) =:= Hpr[A & B & Int]] // <-- doesn't compile

```

## Output

```scala
Cannot prove that Main.AA & Main.BB & Int =:= Hpr[Main.A & Main.B & Int].
        summon[(AA & BB & Int) =:= Hpr[A & B & Int]]
```

## Expectation

Expect "t" to be derived to Int

**Comment by heaven-born** [comment]
This is also looks like an bug:

```scala
type Hpr[X] = X match
     case C => CC // <-- remove this like to make it works
     case A & B => AA & BB

summon[(AA & BB) =:= Hpr[A & B]] // compile error

```

**Comment by odersky** [comment]
Match type reduction is not so intuitive with intersections. To see more it's best to have a setup like in the following test:
```scala
trait A
trait B
trait C
final class D

trait AA
trait BB

type T1[X] = X match
  case A & B => AA & BB

type T2[X] = X match
  case A & t => AA & t

type T3[X] = X match
  case A & B & t => AA & BB & t

val x1 = ??? : T1[A & B]
val x2 = ??? : T2[A & B]
val x3 = ??? : T3[A & B & C]

val y1 = ??? : T1[A & D]
val y2 = ??? : T2[A & D]
val y3 = ??? : T3[A & B & D]
```
If we compile this with -Xprint:typer, we get:
```scala
    val x1: AA & BB = ??? :T1[A & B]
    val x2: AA & (A & B) = ??? :T2[A & B]
    val x3: AA & BB & (A & B & C) = ??? :T3[A & B & C]
    val y1: T1[A & D] = ??? :T1[A & D]
    val y2: T2[A & D] = ??? :T2[A & D]
    val y3: T3[A & B & D] = ??? :T3[A & B & D]
```
We see that the first three lines reduce the match type, but the last three ones leave it as it is, which means the match could not be reduced. There are two particularities here:

 1. when matching with an intersection and a typevar, such as `A & t` in case `x2`, the typevar is always instantiated to the full type of the selector. I.e. `t` is instantiated to `A & B` where it could also be `B`. This would probably be not so easy to fix, intersection types are slippery since they allow many potential solutions. 

 2. When a participant in an intersection type is a final class, there is no match at all. This is actually intentional. Intersections with final classes have no solutions except `Nothing`, so asking a match reduction of `T2[A & D]` is the same as asking `T2[Nothing]`. 
 
The behavior with the added case `case C => CC` also makes sense. The compiler cannot prove that `A & B` is disjoint from `C` since they are all traits and there could well be a class that implements all three. So it cannot simply skip this case to get to the next one. This means it is stuck and there is no reduction.

I leave the issue open since it would be good to document these behaviors.

**Comment by OlivierBlanvillain** [comment]
I agree that the typevar instantiation in `T2[A & B]` and `T3[A & B & C]` is rather surprising, but everything else is already covered by [the match type documentation](https://dotty.epfl.ch/docs/reference/new-types/match-types.html#match-type-reduction
).

In particular, the following sentence explains why the compiler refuses to reduce `Hpr[A & B & Int]` from the original example:

> If the scrutinee type S is an empty set of values (such as Nothing or String & Int), do not reduce.
